[IMP] web_editor,website: enhance snippet search input UX

shsa-odoo opened the pull request:
Steps to reproduce:

1. Click on 'Edit Website'.
2. Observe that the focus is not in snippet search input.
3. Switch tabs from 'Block' to another and vice versa. Two things
   should happen:
   a. Focus should return to the snippet search input.
   b. The search input should be cleared.

Purpose:
    
The purpose of this PR is to save an extra step by automatically
setting focus to the snippet search input when user stars edit mode
previously we have to do it manually by doing something like:
Edit > Click on the search input > Type the name of the block.

Also it handles two more scenarios :

1. Focus is set on the snippet search input when switching from any
  tab to the 'Block' tab in the snippet menu.
2. Any text in the search input is cleared when the user switches
  tabs.

task-3775473

qsm-odoo commented:
@robodoo delegate=bvr-odoo 
@bvr-odoo

shsa-odoo commented:
Please excuse the click that made this PR ready, i looked into the build tour is breaking. will fix it and mark this PR ready once its done.

bvr-odoo commented:
> Please excuse the click that made this PR ready, i looked into the build tour is breaking. will fix it and mark this PR ready once its done.

it's ok, the runbot fail is not related to your PR

bvr-odoo commented:
> Please excuse the click that made this PR ready, i looked into the build tour is breaking. will fix it and mark this PR ready once its done.

It's strange because the first build was ok, the error was not linked to this PR. But after rebuilding it, it is now a fail in your new tour..

bvr-odoo commented on addons/website/static/src/systray_items/edit_website.js:
```diff
@@ -23,6 +23,9 @@ class EditWebsiteSystray extends Component {
                 this.state.isLoading = false;
+                // Once the sinppet is loaded, set the focus on the search input.
+                this.snippetSearchInput = document.querySelector('.o_snippet_search_filter_input');
+                this.snippetSearchInput.focus();
```
As discussed on the task of this PR, this code should be in `web_editor` as we want the same behavior in `Email Marketing`, as the same way it was done for empty the input when switching tabs.

msh-odoo replied:
OK, Agree, initially it was decided to do changes only for website, as we are going keep behavior same, web_editor make sense.

bvr-odoo reviewed (changes requested):
Thanks for your work @shsa-odoo 
Here is my review ;)
As discussed in the task of this PR, the behavior should be the same in `Website` and in `Email Marketing`.
Also, change your commit title, it is not a FIX it is an IMP ;)
Note: I haven't yet reviewed the tour as there will be some changes.

msh-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -3641,6 +3641,15 @@ var SnippetsMenu = Widget.extend({
+            searchInput.focus();
+        }
+        // Clear the search input when switching to the options tab.
+        if (searchInput.value !== '') {
```
OK, we will keep it in mind for forward-port in master

shsa-odoo replied:
Agreed ! I suggest we can have a common method for both the case i.e `_onSnippetSearchResetClick` => `_snippetSearchReset`, i know it would require me to change in xml as well, but having two methods doing the same thing is not a great idea. i will place a comment on this line.

shsa-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -3641,6 +3641,15 @@ var SnippetsMenu = Widget.extend({
         }
 
         this.$('.o_snippet_search_filter').toggleClass('d-none', this._currentTab !== this.tabs.BLOCKS);
+        const searchInput = this.el.querySelector('.o_snippet_search_filter_input');
```
@bvr-odoo I will keep those things in mind in all the PR's to come.Thanks for the review! :)

shsa-odoo commented on addons/website/static/src/systray_items/edit_website.js:
```diff
@@ -23,6 +23,9 @@ class EditWebsiteSystray extends Component {
         useEffect((snippetsLoaded) => {
             if (snippetsLoaded) {
                 this.state.isLoading = false;
+                // Once the sinppet is loaded, set the focus on the search input.
```
my bad :sweat_smile:

shsa-odoo commented on addons/website/static/src/systray_items/edit_website.js:
```diff
@@ -23,6 +23,9 @@ class EditWebsiteSystray extends Component {
             if (snippetsLoaded) {
                 this.state.isLoading = false;
+                // Once the sinppet is loaded, set the focus on the search input.
+                this.snippetSearchInput = document.querySelector('.o_snippet_search_filter_input');
```
@bvr-odoo Its better i should do  `document.querySelector('.o_snippet_search_filter_input'),focus()`, instead of adding two line, i will change that. Thanks !

bvr-odoo commented:
@shsa-odoo Could you change the title of the commit as I asked [here] please ;) (https://github.com/odoo/odoo/pull/158033#pullrequestreview-2011370238)
Also, we discussed about this task here in the team and we think it is better to rebase it in Master. Indeed, it is not a big IMP so it is safer to target Master. S ocould you rebase and change the title please, thanks a lot.

bvr-odoo reviewed (changes requested):
@shsa-odoo I converted your code in OWL (no longer need for the `setTimeout`, etc.). Next time try to use OWL when it's possible, thanks ;)

There is still something wrong on `mass_mailing` ...
When I have something selected in the template (text or snippet) see the picutre:
![image](https://github.com/odoo/odoo/assets/52911687/bddcef11-0ae8-436a-ac93-74d92758099a)
And then if I click on one of the buttons, inputs on the "o_mass_mailing_mailing_group" top section (Subject, Recipients, etc.), the search input is focused and Selected. If I want to write a subject, for example, I now have to click twice on the input (instead of only one click).
I guess to code you removed is still needed, could you check that please, thanks.

shsa-odoo commented:
> @shsa-odoo I converted your code in OWL (no longer need for the `setTimeout`, etc.). Next time try to use OWL when it's possible, thanks ;)
> 
> There is still something wrong on `mass_mailing` ... When I have something selected in the template (text or snippet) see the picutre: ![image](https://private-user-images.githubusercontent.com/52911687/325234317-bddcef11-0ae8-436a-ac93-74d92758099a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwMzU3MzYsIm5iZiI6MTcxNDAzNTQzNiwicGF0aCI6Ii81MjkxMTY4Ny8zMjUyMzQzMTctYmRkY2VmMTEtMGFlOC00MzZhLWFjOTMtNzRkOTI3NTgwOTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI1VDA4NTcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhMTk2NjcyMzljNThiYTljMjE4YWZlNDM4ZTU4YzlhZGZhMmEzZjlmODk2N2ZjZDU5ZWMwMjc3Yzc4NGU1ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Se_PRKa0xkOs_KASwcgF-gcskvIvfnoKg_4ssAVe3u8) And then if I click on one of the buttons, inputs on the "o_mass_mailing_mailing_group" top section (Subject, Recipients, etc.), the search input is focused and Selected. If I want to write a subject, for example, I now have to click twice on the input (instead of only one click). I guess to code you removed is still needed, could you check that please, thanks.

@bvr-odoo Definitely, i will look into the issue and ping you once done, Thanks !

shsa-odoo commented:
> Will mention in passing, there's something weird going on in mailing where it briefly flashes the regular snippet menu when you open a new mailing. It's a bit jarring, would be nice if avoidable.

@reth-odoo IMO that could be happening because the snippet editor mount promise might not have resolved by the time snippet menu component is mounted.

shsa-odoo commented:
> @shsa-odoo I converted your code in OWL (no longer need for the `setTimeout`, etc.). Next time try to use OWL when it's possible, thanks ;)
> 
> There is still something wrong on `mass_mailing` ... When I have something selected in the template (text or snippet) see the picutre: ![image](https://private-user-images.githubusercontent.com/52911687/325234317-bddcef11-0ae8-436a-ac93-74d92758099a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQxMTMxNTMsIm5iZiI6MTcxNDExMjg1MywicGF0aCI6Ii81MjkxMTY4Ny8zMjUyMzQzMTctYmRkY2VmMTEtMGFlOC00MzZhLWFjOTMtNzRkOTI3NTgwOTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI2VDA2MjczM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVmZGVmODkwZWNiYzEyMTJhMDM4MGNmYWM2ZjY0NTMwYTAxMjhjOWQ1ZjczZDUxNTRlNzVhYjBlNjU3ZWU1YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FDbi9G82jbqFs1BIb4QxLG4O7chE0wjhi7ZHi5Bo7ks) And then if I click on one of the buttons, inputs on the "o_mass_mailing_mailing_group" top section (Subject, Recipients, etc.), the search input is focused and Selected. If I want to write a subject, for example, I now have to click twice on the input (instead of only one click). I guess to code you removed is still needed, could you check that please, thanks.

Hello @bvr-odoo,i looked into the issue, the focus is not set on the input search for mass mailing, reason being the `useEffect` doesnot gets triggered initially, when the snippet loads, Also the problem we see in your screenshot, is not because of the code i removed, that was there to set the cursor to the editable, the reason it takes two click is that, is when we click away in that input field of 'o_mass_mailing_mailing_group' the dependency we gave to 'useEffect` in our case [this.state.currenTab], changes and our method for useEffect gets triggered, which forces our code to set focus back to input search field, and when we click again since there is no state change the focus gets to clicked input field.Thanks!

shsa-odoo commented:
@bvr-odoo in resposne to make the focus set in input search for the case of `mass_mailing`, we will somewhat do it in [here](https://github.com/odoo/odoo/blob/17.0/addons/mass_mailing/static/src/js/mass_mailing_html_field.js#L382), which loads our `oe_snippet` after  animation frame, and again solution to our problem which requires double click in input of `o_mass_mailing_mailing_group` we have to add an event listener for click event after theme is selected [here](https://github.com/odoo/odoo/blob/master/addons/mass_mailing/static/src/js/snippets.editor.js#L31), which i wonder should be a good approach or the solution we want, IMO, we should consider the initial approach i used to solve this issue in `-updateRightPanelContent`, after animation frame, what is your view on this  ?

bvr-odoo commented:
@shsa-odoo

![image](https://github.com/odoo/odoo/assets/52911687/4ca745e2-c27b-4d5b-91ec-2195cf94082d)

After thinking about it, I believe we should do what PDE wanted, which is not to touch `mass_mailing`. The behavior is indeed different from Website because we start editing a template with existing content on the template. So, make sure everything works in Website and leave `mass_mailing` as it is currently. I know this wasn't what I originally asked for, but now that I understand the different behaviors better, I think we should do what PDE requested. I hope this will make things simpler for this PR.

bvr-odoo reviewed (changes requested):
This PR should have not impact the `mass_mailling` app as discussed [here](https://github.com/odoo/odoo/pull/158033#issuecomment-2084730911)
This issue still exist => 
https://github.com/odoo/odoo/pull/158033#issuecomment-2075025263
And this one too => 
https://github.com/odoo/odoo/pull/158033#pullrequestreview-2019863071

shsa-odoo commented:
> This PR should have not impact the `mass_mailling` app as discussed [here](https://github.com/odoo/odoo/pull/158033#issuecomment-2084730911) This issue still exist => [#158033 (comment)](https://github.com/odoo/odoo/pull/158033#issuecomment-2075025263) And this one too => [#158033 (review)](https://github.com/odoo/odoo/pull/158033#pullrequestreview-2019863071)

Hello @bvr-odoo the problem mentioned in [comment](https://github.com/odoo/odoo/pull/158033#issuecomment-2075025263) is all together a different issue, as i mentioned in this [respose](https://github.com/odoo/odoo/pull/158033#issuecomment-2078839792), its totally unrelated to focus issue, and should be fixed inanother PR, i have a solution to that as well, so do you want i should fix in this PR, ?or i make another PR specifically for this issue. 
Meanwhile @msh-odoo the [commit](https://github.com/odoo/odoo/commit/4952aab6a076f4015bd1f222e14dbe972167452c)  I made resolved that issue which @bvr-odoo mentioned on this [comment](https://github.com/odoo/odoo/pull/158033#pullrequestreview-2019863071)  i made all the changes in website module's snippet editor only so that the changes wont affect the mass_mailing module. Making changes back to snippet_editor will bring us to same scenario where we have to manually make preventive code to stop mass_mailing getting affected, and i don't think we would want that, right ? Your views on those changes you made is welcomed, Thanks!

msh-odoo commented:
> > This PR should have not impact the `mass_mailling` app as discussed [here](https://github.com/odoo/odoo/pull/158033#issuecomment-2084730911) This issue still exist => [#158033 (comment)](https://github.com/odoo/odoo/pull/158033#issuecomment-2075025263) And this one too => [#158033 (review)](https://github.com/odoo/odoo/pull/158033#pullrequestreview-2019863071)
> 
> Hello @bvr-odoo the problem mentioned in [comment](https://github.com/odoo/odoo/pull/158033#issuecomment-2075025263) is all together a different issue, as i mentioned in this [respose](https://github.com/odoo/odoo/pull/158033#issuecomment-2078839792), its totally unrelated to focus issue, and should be fixed inanother PR, i have a solution to that as well, so do you want i should fix in this PR, ?or i make another PR specifically for this issue. Meanwhile @msh-odoo the [commit](https://github.com/odoo/odoo/commit/4952aab6a076f4015bd1f222e14dbe972167452c) I made resolved that issue which @bvr-odoo mentioned on this [comment](https://github.com/odoo/odoo/pull/158033#pullrequestreview-2019863071) i made all the changes in website module's snippet editor only so that the changes wont affect the mass_mailing module. Making changes back to snippet_editor of web_editor will bring us to same scenario where we have to manually make preventive code to stop mass_mailing getting affected, and i don't think we would want that, right ? Your views on those changes you made is welcomed so that we can move ahead accordingly, Thanks!

@bvr-odoo It was me who forgot that we have to skip mass_mailing to affect as I thought right place for code is web_editor module but as you said we don't want to touch mass_mailing in that case website editor is right place, @shsa-odoo Please do the changes which you did in website module. Thanks.

qsm-odoo reviewed (approved):
Thanks for the PR, I push-forced to make the final cleaning and fixes: https://github.com/odoo/odoo/compare/6ee4681a8e9bd4df72433bee63679433ac27f526..6673a47b8e792fe25e766b7e3992647a1e751fbc
- Some nitpicking (uppercase letter to start comment, one liners, ...)
- Removed useless comments
- Use of `TourError` in tours. I actually did not know about that one, seems to be the new way to recommend.
- Use of `this.anchor` in tour instead of re-searching the DOM in `run`
- **Important**: removed the use of `isEqualNode` which was just a mistake. `isEqualNode` does not check that this is the same node, it just checks that it looks the same.

@robodoo r+

shsa-odoo commented:
> Thanks for the PR, I push-forced to make the final cleaning and fixes: https://github.com/odoo/odoo/compare/6ee4681a8e9bd4df72433bee63679433ac27f526..6673a47b8e792fe25e766b7e3992647a1e751fbc
> 
> * Some nitpicking (uppercase letter to start comment, one liners, ...)
> * Removed useless comments
> * Use of `TourError` in tours. I actually did not know about that one, seems to be the new way to recommend.
> * Use of `this.anchor` in tour instead of re-searching the DOM in `run`
> * **Important**: removed the use of `isEqualNode` which was just a mistake. `isEqualNode` does not check that this is the same node, it just checks that it looks the same.
> 
> @robodoo r+

@qsm-odoo Noted, will remember this in my next PR, Thanks! :+1:
